### PR TITLE
add tests for Eperience

### PR DESCRIPTION
**Pull request by R-Morpheus**

### CODE REVIEW

File: packages/contracts/test/LibExperience.t.sol [hunk 1]
```diff
@@ -27,4 +28,64 @@ contract LibExperienceTest is MudV2Test {
...
+    assertTrue(LibExperience.hasExp(targetEntity));
+
+    uint32[PStat_length] memory addExp = [uint32(1), 1, 1];
+    LibExperience.increaseExp(targetEntity, addExp);
+
+    uint32[PStat_length] memory exp = LibExperience.getExp(targetEntity);
+    for (uint256 i; i < exp.length; i++) {
+      assertEq(exp[i], 1);
+    }
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:43]: this is just a copy of `testIncreaseExp`, you don't need it twice

File: packages/contracts/test/LibExperience.t.sol [hunk 2]
```diff
@@ -27,4 +28,64 @@ contract LibExperienceTest is MudV2Test {
...
+
+  function testHasExp() public {
+    assertFalse(LibExperience.hasExp(targetEntity));
+
+    LibExperience.initExp(targetEntity);
+    assertTrue(LibExperience.hasExp(targetEntity));
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:50]: you can remove
```solidity
LibExperience.initExp(targetEntity);
assertTrue(LibExperience.hasExp(targetEntity));
```
from all other tests, since you add a dedicated existence check

File: packages/contracts/test/LibExperience.t.sol [hunk 3]
```diff
@@ -27,4 +28,64 @@ contract LibExperienceTest is MudV2Test {
...
+
+  function testMaxLevel() public {
+    LibExperience.initExp(targetEntity);
+
+    // Set the experience to the maximum level
+    uint32[PStat_length] memory maxExp = [type(uint32).max, type(uint32).max, type(uint32).max];
+    Experience.set(targetEntity, maxExp);
+
+    assertEq(LibExperience.getAggregateLevel(targetEntity, maxExp), 16);
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:90]: this just tests that an exp cap exists,
you could also test for specific pstat values using `getExpForPStats`,
fuzz tests for pstat values might also be nice

File: packages/contracts/test/LibExperience.t.sol [hunk 4]
```diff
@@ -27,4 +28,64 @@ contract LibExperienceTest is MudV2Test {
...
+    uint32[PStat_length] memory expectedExp;
+    for (uint256 i; i < initialExp.length; i++) {
+      expectedExp[i] = initialExp[i] + addExp[i];
+    }
+
+    uint32[PStat_length] memory retrievedExp = LibExperience.getExp(targetEntity);
+    for (uint256 i; i < retrievedExp.length; i++) {
+      assertEq(retrievedExp[i], expectedExp[i]);
+    }
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:80]: this also mostly duplicates `testIncreaseExp` (but adds an unnecessary loop),
instead you could make this a `testFuzzIncreaseExp`, see https://book.getfoundry.sh/forge/fuzz-testing

**Review by dk1a — COMMENTED** [review]
spacing was still a mess, I added a staged-linter so u don't have to deal with it.
remove `package-lock.json`, it's created by `npm`, but u should use `pnpm`.
I also fixed `testMaxLevel`, see the commit.

File: packages/contracts/test/LibExperience.t.sol [hunk 5]
```diff
@@ -7,85 +7,68 @@ import { Experience, ExperienceTableId } from "../src/codegen/Tables.sol";
...
-  function setUp() public virtual override {
-    super.setUp();
-    // TODO remove this if it's integrated into MUD
-    vm.startPrank(worldAddress);
-  }
+    function setUp() public virtual override {
+        super.setUp();
+        // TODO remove this if it's integrated into MUD
+        vm.startPrank(worldAddress);
+    }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:16]: change ur linter to use 2 spaces

File: packages/contracts/test/LibExperience.t.sol [hunk 6]
```diff
@@ -24,7 +25,50 @@ contract LibExperienceTest is MudV2Test {
 
     uint32[PStat_length] memory exp = LibExperience.getExp(targetEntity);
     for (uint256 i; i < exp.length; i++) {
-      assertEq(exp[i], 1);
+        assertEq(exp[i], 1);
     }
   }
+
+  function testHasExp() public {
+    assertFalse(LibExperience.hasExp(targetEntity));
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:33]: why change `testHasExp`? it was the only test I liked from the start

File: packages/contracts/test/LibExperience.t.sol [hunk 7]
```diff
@@ -24,7 +25,50 @@ contract LibExperienceTest is MudV2Test {
 
     uint32[PStat_length] memory exp = LibExperience.getExp(targetEntity);
     for (uint256 i; i < exp.length; i++) {
-      assertEq(exp[i], 1);
+        assertEq(exp[i], 1);
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:28]: run `pnpm install` to set up staged linter

File: packages/contracts/test/LibExperience.t.sol [hunk 8]
```diff
@@ -24,7 +25,50 @@ contract LibExperienceTest is MudV2Test {
...
+
+  function testSetExp() public {
+    uint32[PStat_length] memory exp = [uint32(10), 20, 30];
+    Experience.set(targetEntity, exp);
+
+    uint32[PStat_length] memory retrievedExp = LibExperience.getExp(targetEntity);
+    for (uint256 i; i < retrievedExp.length; i++) {
+        assertEq(retrievedExp[i], exp[i]);
+    }
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:44]: naming nit: this should be `testGetExp`
you're not really testing `Experience.set`, since tables're tested by MUD internally. `getExp` is a custom library function however

File: packages/contracts/src/guise/LibGuiseLevel.sol [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import { Experience, ExperienceTableId } from "../codegen/Tables.sol";
+import { ActiveGuise, ActiveGuiseTableId } from "../codegen/Tables.sol";
+import { GuisePrototype, GuisePrototypeTableId } from "../codegen/Tables.sol";
+
+import { PStat, PStat_length } from "../CustomTypes.sol";
+import { LibExperience } from "../charstat/LibExperience.sol";
+
+library LibGuiseLevel {
+  using LibExperience for *;
```
Comment (dk1a) [inline packages/contracts/src/guise/LibGuiseLevel.sol:15]: In general you shouldn't use `*` in `using for` (I didn't even know you could),
and in this case I don't think `LibExperience` is a good fit for `using for`, since not all of its functions have the same kind of first arg

File: packages/contracts/src/guise/LibGuiseLevel.sol [hunk 10]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.17;
+
+import { IUint256Component } from "@latticexyz/solecs/src/interfaces/IUint256Component.sol";
```
Comment (dk1a) [inline packages/contracts/src/guise/LibGuiseLevel.sol:5]: this file doesn't work and looks unfinished
in particular both pragma and the old import are invalid, I assume u didn't test this before pushing?

**Review by dk1a — COMMENTED** [review]
please check ur commits to make sure no extra files are present, and that the files compile and tests pass at least
you can add both `.idea` and `.DS_Store` to `.gitignore`

File: packages/contracts/test/LibExperience.t.sol (hunk 6 above)
Comment (R-Morpheus) [reply packages/contracts/test/LibExperience.t.sol:33]: I'm not sure here, before you indicated to delete the extra initialization lines
```
    LibExperience.initExp(targetEntity);
    assertTrue(LibExperience.hasExp(targetEntity));
```

Comment (dk1a) [reply packages/contracts/test/LibExperience.t.sol:33]: > you can remove
> ```solidity
> LibExperience.initExp(targetEntity);
> assertTrue(LibExperience.hasExp(targetEntity));
> ```
> from all other tests, since you add a dedicated existence check

I meant that `testHasExp` does an existence check, so other tests don't need it

File: packages/contracts/.gitignore [hunk 11]
```diff
@@ -6,6 +6,8 @@ artifacts/
 abi/
 types/
 broadcast/
+.idea/
+.DS_Store
```
Comment (dk1a) [inline packages/contracts/.gitignore:10]: you should use the root `.gitignore`, this one only affects the `contracts` folder.
and you should also remove the files themselves

File: packages/contracts/src/guise/LibGuiseLevel.sol [hunk 12]
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { ActiveGuise, ActiveGuiseTableId } from "../codegen/Tables.sol";
+import { GuisePrototype, GuisePrototypeTableId } from "../codegen/Tables.sol";
```
Comment (dk1a) [inline packages/contracts/src/guise/LibGuiseLevel.sol:5]: this can be a single import

File: packages/contracts/src/guise/LibGuiseLevel.sol [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+import { PStat_length } from "../CustomTypes.sol";
+import { LibExperience } from "../charstat/LibExperience.sol";
+
+library LibGuiseLevel {
+  /// @dev Get target's aggregate level using its guise's level multipliers.
+  /// (aggregate level means all primary stats aggregated together)
+  function getAggregateLevel(bytes32 targetEntity) internal view returns (uint32) {
+    uint256 guiseProtoEntity = ActiveGuise.get(targetEntity);
+    uint32[PStat_length] memory levelMul = GuisePrototype.get(bytes32(guiseProtoEntity));
```
Comment (dk1a) [inline packages/contracts/src/guise/LibGuiseLevel.sol:15]: the typecast is unnecessary

File: packages/contracts/src/guise/LibGuiseLevel.sol [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    return LibExperience.getAggregateLevel(targetEntity, levelMul);
+  }
+
+  /// @dev Multiply gained experience by guise's level multiplier
+  function multiplyExperience(
+    bytes32 targetEntity,
+    uint32[PStat_length] memory exp
+  ) internal view returns (uint32[PStat_length] memory expMultiplied) {
+    uint256 guiseProtoEntity = ActiveGuise.get(targetEntity);
+    uint32[PStat_length] memory levelMul = GuisePrototype.get(bytes32(guiseProtoEntity));
```
Comment (dk1a) [inline packages/contracts/src/guise/LibGuiseLevel.sol:25]: same here

File: packages/contracts/test/LibExperience.t.sol [hunk 15]
```diff
@@ -27,4 +33,56 @@ contract LibExperienceTest is MudV2Test {
...
+    uint32[PStat_length] memory expectedExp = LibExperience.getExpForPStats(pstats);
+
+    // Get the actual exp
+    uint32[PStat_length] memory actualExp = LibExperience.getExp(targetEntity);
+
+    // Ensure that the actual exp matches the expected exp
+    for (uint256 i = 0; i < expectedExp.length; i++) {
+      assert(actualExp[i] == expectedExp[i]);
+    }
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:87]: 1. only functions that start with `test` get run. `fuzz` doesn't have any magic meaning. `testFuzzPStats` or `test_fuzz_PStats` would work
2. exp is neither initialized, nor set
3. never use `assert`, it doesn't do what you think it does. Here `assertEq` would work
4. these functions should be `public`

File: packages/contracts/test/LibExperience.t.sol [hunk 16]
```diff
@@ -27,4 +33,56 @@ contract LibExperienceTest is MudV2Test {
...
+    uint32[PStat_length] memory expectedExp;
+    for (uint256 i; i < initialExp.length; i++) {
+      expectedExp[i] = initialExp[i] + addExp[i];
+    }
+
+    uint32[PStat_length] memory retrievedExp = LibExperience.getExp(targetEntity);
+    for (uint256 i; i < retrievedExp.length; i++) {
+      assertEq(retrievedExp[i], expectedExp[i]);
+    }
+  }
```
Comment (dk1a) [inline packages/contracts/test/LibExperience.t.sol:64]: this is still just like `testIncreaseExp`, but looks needlessly complicated and with slightly different values

File: packages/contracts/test/LibGuiseLevel.t.sol [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+  function testGetAggregateLevel() public {
+    // Initialize exp and levelMul
+    LibExperience.initExp(targetEntity);
+    assertTrue(LibExperience.hasExp(targetEntity));
+
+    // Add exp
+    uint32[PStat_length] memory addExp = [uint32(1), 1, 1];
+    LibExperience.increaseExp(targetEntity, addExp);
+
+    //Add guise and set levelmul
```
Comment (dk1a) [inline packages/contracts/test/LibGuiseLevel.t.sol:29]:
```suggestion
    // Add guise and set levelmul
```

File: packages/contracts/test/LibGuiseLevel.t.sol [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    uint256 guiseProtoEntity = ActiveGuise.get(targetEntity);
+    uint32[PStat_length] memory levelMul = [uint32(2), 2, 2];
+    GuisePrototype.set(bytes32(guiseProtoEntity), levelMul);
+
+    uint32 aggregateLevel = LibGuiseLevel.getAggregateLevel(bytes32(targetEntity));
+    assertEq(aggregateLevel, 1); // expected 1 = (2*1 + 2*1 + 2*1) / (2 + 2 + 2)
```
Comment (dk1a) [inline packages/contracts/test/LibGuiseLevel.t.sol:35]: for a one-off test I'd prefer more complicated values.
But better yet you should make a helper function instead and test several sets of edge cases.
And then you could fuzz it too

File: packages/contracts/test/LibGuiseLevel.t.sol [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+  function setUp() public virtual override {
+    super.setUp();
+    vm.startPrank(worldAddress);
+  }
+
+  function testGetAggregateLevel() public {
+    // Initialize exp and levelMul
+    LibExperience.initExp(targetEntity);
+    assertTrue(LibExperience.hasExp(targetEntity));
```
Comment (dk1a) [inline packages/contracts/test/LibGuiseLevel.t.sol:23]: I think it's fine to put this into `setUp`, since you test exp in `LibExperienceTest`

File: packages/contracts/test/LibGuiseLevel.t.sol [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+  function testMultiplyExperience() public {
+    LibExperience.initExp(targetEntity);
+    assertTrue(LibExperience.hasExp(targetEntity));
+
+    uint32[PStat_length] memory addExp = [uint32(1), 1, 1];
+    LibExperience.increaseExp(targetEntity, addExp);
+
+    uint256 guiseProtoEntity = ActiveGuise.get(targetEntity);
+    uint32[PStat_length] memory levelMul = [uint32(2), 2, 2];
+    GuisePrototype.set(bytes32(guiseProtoEntity), levelMul);
```
Comment (dk1a) [inline packages/contracts/test/LibGuiseLevel.t.sol:47]: this is repeated verbatim from `testGetAggregateLevel`, so they could share the same helper function

File: packages/contracts/src/guise/LibGuiseLevel.sol (hunk 13 above)
Comment (R-Morpheus) [reply packages/contracts/src/guise/LibGuiseLevel.sol:15]: I tried to write without type casting, but with `pnpm run test`, the compiler reported an error of the arguments being passed, since `GuisePrototype.get` accepts `bytes32`, and ActiveGuise.get returns `uint256`
Снимок экрана 2023-06-05 в 16 11 37

Comment (dk1a) [reply packages/contracts/src/guise/LibGuiseLevel.sol:15]: oh right I changed the tables to `bytes32` in https://github.com/dk1a/wanderer-cycle2/pull/2
